os/arch/arm/src/imxrt/imxrt_timer_lowerhalf.c

```diff
@@ -0,0 +1,457 @@
+ * Name: imxrt_gpt_ioctl
+ *
+ * Description:
+ *   Any ioctl commands that are not recognized by the "upper-half" driver
+ *   are forwarded to the lower half driver through this method.
+ *
+ * Input Parameters:
+ *   lower - A pointer the publicly visible representation of the "lower-half"
+ *           driver state structure.
+ *   cmd   - The ioctl command value
+ *   arg   - The optional argument that accompanies the 'cmd'.  The
+ *           interpretation of this argument depends on the particular
+ *           command.
+ *
+ * Returned Values:
+ *   Zero on success; a negated errno value on failure.
+ *
+ ****************************************************************************/
+static int imxrt_gpt_ioctl(struct timer_lowerhalf_s *lower, int cmd,
+						unsigned long arg)
+{
+	struct imxrt_gpt_lowerhalf_s *priv = (struct imxrt_gpt_lowerhalf_s *)lower;
+	int ret = ERROR;
+	gpt_clock_source_t clock_source = priv->config.clockSource;
+
+	tmrvdbg("GPT IOCTL cmd: %d, arg: %ld\n", cmd, arg);
+
+	switch (cmd) {
+	case TCIOC_SETDIV:
+		if (clock_source == kGPT_ClockSource_Osc) {
+			if (arg <= GPT_OSCDIV_MAXVAL) {
+				imxrt_gpt_setoscclockdivider(priv->gpt->base, arg);
+				priv->config.divider = arg;
+				ret = OK;
+			} else {
+				tmrdbg("Invalid Value of OSC Divider (%d)\n", arg);
+			}
+		} else {
+			if (arg <= GPT_DIV_MAXVAL) {
+				imxrt_gpt_setclockdivider(priv->gpt->base, arg);
+				priv->config.divider = arg;
+				ret = OK;
+			} else {
+				tmrdbg("Invalid Value of Divider (%d)\n", arg);
+			}
+		}
+		break;
+	case TCIOC_GETDIV:
+		if (clock_source == kGPT_ClockSource_Osc) {
+			ret = (int)imxrt_gpt_getoscclockdivider(priv->gpt->base);
+		} else {
+			ret = (int)imxrt_gpt_getclockdivider(priv->gpt->base);
+		}
+		break;
+	default:
+		tmrdbg("Invalid cmd %d\n", cmd);
+		break;
+	}
+
+	return ret;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+int imxrt_timer_initialize(const char *devpath, int timer)
+{
+	struct imxrt_gpt_lowerhalf_s *lower = NULL;
+	gpt_config_t config;
+
+	switch (timer) {
+	case IMXRT_GPT_CH0:
+		lower = &g_gpt0_lowerhalf;
+		break;
+
+	case IMXRT_GPT_CH1:
+		lower = &g_gpt1_lowerhalf;
+		break;
+	}
+
+	if (!lower) {
+		return -ENODEV;
+	}
+
+	/* Initialize the elements of lower half state structure */
+	lower->started  = false;
+	lower->callback = NULL;
+
+	imxrt_gpt_getdefaultconfig(&config);
+	memcpy(&(lower->config), &config, sizeof(gpt_config_t));
+	lower->gpt = imxrt_gpt_configure(timer, &config);
+
+	if (lower->gpt == NULL) {
+		return -EINVAL;
+	}
+
+	/*
+	 * Register the timer driver as /dev/timerX.  The returned value from
+	 * timer_register is a handle that could be used with timer_unregister().
+	 * REVISIT: The returned handle is discard here.
+	 */
+	if (!timer_register(devpath, (struct timer_lowerhalf_s *)lower)) {
```

Comment:
This is lowerhalf of timer, so it should be registered to Timer as a interface of lower device controller.
So timer_register should be called in board code, I think imxrt_boot.c is acceptable.
Board code should init config first, and then init lowerhalf with config, finally timer.
I don't know about timer properly, but it seems that below should be applied.
```
struct imxrt_gpt_lowerhalf_s *gpt
gpt = imxrt_gpt_configure();
if (gpt == NULL) {
   error;
}
timer_lowerhalf_s *timer = imxrt_timer_initialize(gpt );
if (timer == NULL) {
  error;
}
timer_register(devpath, timer);
....
```